### DISCUSSION TITLE
UserData nesting for external types

### DISCUSSION THREAD

**Discussion by IcyTv** [mlua-rs/mlua#186] [Q&A]
Hi, I'm currently having the following problem: I have external dependent types, something like:
```rust
struct Vec3 {}
struct Mat3 {
    x_axis: Vec3;
    y_axis: Vec3;
    z_axis: Vec3;
}
```
from an external crate.
I now want to bind Mat3 into the Lua context. To do this I (have to) use the rust "newtype" pattern, so:
```rust
struct LuaVec3(Vec3);
impl UserData for LuaVec3 {...}
struct LuaMat3(Mat3);
impl UserData for LuaMat3 {...}
```
and implement the `UserData` for those.
If I now, however, write an accessor for one of the axis for Lua, I have to wrap the resulting Value in my crate-local files. This leads (because of the ownership) to a copied Object in Lua and if I edit that, the original axis (obviously) doesn't get changed...

The only 2 options I see to circumvent this is to, either reimplement these types in my crate (not something I'm very keen on) or modify the original (glam) crate to have all the external types implement `UserData`.
Is there something I'm missing, i.e. a way to circumvent the orphan rules that gives me the ability to retain ownership of the object with `Mat3`? Or implementing `UserData` for the references? Or have a reference wrapper (how would that work?)?

Your help would be much appreciated.

**Comment by dubrowgn** [comment]
For anyone still looking for an answer to this like I was, I found a relatively straight forward way to do this for non-`UserData` types:

```rust
// main.rs -- assume Name is an external type

pub struct Name {
	pub name: String,
}

impl Name {
	pub const fn new(name: String) -> Self { Self { name } }
}

impl fmt::Debug for Name {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { self.name.fmt(f) }
}


fn main() -> Result<(), Box<dyn std::error::Error>>  {
	let lua = Lua::new();

	// calling `register_userdata_type::<T>(...)` is roughly equivalent to `impl UserData for T`

	lua.register_userdata_type::<Name>(|reg| {
		reg.add_field_method_get("name", |lua, this| Ok(this.name.as_str().into_lua(lua)?));
		reg.add_field_method_set("name", |_, this, val| { this.name = val; Ok(()) });
		reg.add_meta_function(LuaMetaMethod::ToString, |_, this: LuaUserDataRef<Name>| Ok(format!("{:?}", this)));
	})?;

	lua.globals().set("Name", lua.create_function(|lua, name| {
		let n = if let Some(name) = name { Name::new(name) } else { Name::default() };
		lua.create_any_userdata(n)
	})?)?;

	Ok(())
}
```

```lua
-- script.lua

local n1 = Name("foo");
print("typeof(n1):", n1);
print("n1:", n1);
print("n1.name:", n1.name);

local n2 = Name();
print("n2:", n2);
n2.name = "bar";
print("n2:", n2);
```

Results in:

```txt
typeof(n1):	Name
n1:	"foo"
n1.name:	"foo"
n2:	""
n2:	"bar"
```

**Reply by kpfleming** [reply]
Thank you, this is very helpful!

**Reply by kpfleming** [reply]
To go back to your original example, if you've used this method to register a UserData type for `Vec3`, how do you setup the field methods for `Mat` to provide access to its fields?

**Reply by dubrowgn** [reply]
`register_userdata_type` solves the problem of implementing `UserData` for external types, but unfortunately I don't know how to return a reference to a field to Lua. This isn't something I've needed so far.

Maybe you could do something with `LuaUserDataRef`, but you might have to just change your Lua API to avoid direct references to the axis fields.

**Reply by kpfleming** [reply]
From what I've been able to tell Rust doesn't even have 'pointers to fields' (which are pointers-to-member in C++), so it's never going to be possible to do that.

The options appear to be:

1. Provide field `get` and `set` on the type which allow the user to obtain and replace the contents of the field. In your example this would work fine since `Vec3` is a cheaply copyable type.

2. Ensure that the type of the field itself is cheaply clonable, even if it would not normally be. In most cases this will involve wrapping it in `Rc<RefCell<...>>` or `Arc<Mutex<...>>` but other arrangements could work. Using the registry to provide UserData makes this easier as you should be able to directly register UserData for `Rc<RefCell<Vec3>>` and not rely on the `userdata-wrappers` feature to do that for you. The field `get` would then `clone` the field's contents - the result is a pointer to the data that field points to, but not a pointer to the field itself (if the field is changed to point to different contents, the value previously returned to Lua from `get` would still point to the original contents).

3. Implement methods (not fields) to retrieve and manipulate the contents of the field.
